<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="fr" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Introduction to data analysis with R and Bioconductor: Manipulation et analyse de donn√©es avec dplyr</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png"><link rel="manifest" href="../favicons/incubator/site.webmanifest"><link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"></div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../30-dplyr.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Introduction to data analysis with R and Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Introduction to data analysis with R and Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Introduction to data analysis with R and Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 43%" class="percentage">
    43%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 43%" aria-valuenow="43" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../30-dplyr.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="10-data-organisation.html">1. Organisation des donn√©es avec des feuilles de calcul</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="20-r-rstudio.html">2. R et RStudio</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="23-starting-with-r.html">3. Introduction √† R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="25-starting-with-data.html">4. Commencer par les donn√©es</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        5. Manipulation et analyse de donn√©es avec dplyr
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#manipulation-des-donn%C3%A9es-avec-dplyr-et-tidyr">Manipulation des donn√©es avec <strong><code>dplyr</code></strong> et
<strong><code>tidyr</code></strong></a></li>
<li><a href="#chargement-de-donn%C3%A9es-avec-le-tidyverse">Chargement de donn√©es avec le tidyverse</a></li>
<li><a href="#s%C3%A9lection-de-colonnes-et-filtrage-de-lignes">S√©lection de colonnes et filtrage de lignes</a></li>
<li><a href="#pipes">Pipes</a></li>
<li><a href="#mutate">Mutate</a></li>
<li><a href="#analyse-de-donn%C3%A9es-split-apply-combine">Analyse de donn√©es ‚Äúsplit-apply-combine‚Äù</a></li>
<li><a href="#remodeler-les-donn%C3%A9es">Remodeler les donn√©es</a></li>
<li><a href="#jointures-de-tables">Jointures de tables</a></li>
<li><a href="#exporter-des-donn%C3%A9es">Exporter des donn√©es</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="40-visualization.html">6. Visualisation de donn√©es</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="60-next-steps.html">7. Prochaines √©tapes</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/25-starting-with-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/40-visualization.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/25-starting-with-data.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Commencer par les
        </a>
        <a class="chapter-link float-end" href="../instructor/40-visualization.html" rel="next">
          Next: Visualisation de...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Manipulation et analyse de donn√©es avec dplyr</h1>
        <p>Last updated on 2025-05-05 |

        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/episodes/30-dplyr.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 150 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>Analyse de donn√©es en R √† l‚Äôaide du m√©ta-package (paquet)
tidyverse</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>D√©crire l‚Äôobjectif des packages <strong><code>dplyr</code></strong>
et <strong><code>tidyr</code></strong>.</li>
<li>D√©crire plusieurs fonctions extr√™mement utiles pour manipuler des
donn√©es.</li>
<li>D√©crire le concept d‚Äôun tableau large et long, et comment remodeler
un tableau d‚Äôun format √† l‚Äôautre.</li>
<li>Montrer comment joindre des tables.</li>
</ul></div>
</div>
</div>
</div>
</div>
<blockquote>
<p>Cet √©pisode est bas√© sur la le√ßon <em>Data Analysis and Visualisation
in R for Ecologists</em> de Data Carpentries.</p>
</blockquote>
<section><h2 class="section-heading" id="manipulation-des-donn√©es-avec-dplyr-et-tidyr">Manipulation des donn√©es avec <strong><code>dplyr</code></strong> et
<strong><code>tidyr</code></strong>
<a class="anchor" aria-label="anchor" href="#manipulation-des-donn%C3%A9es-avec-dplyr-et-tidyr"></a></h2>
<hr class="half-width"><p>Extraire un sous-ensemble avec les crochets est pratique, mais peut
√™tre fastidieux et difficile √† lire, en particulier pour les op√©rations
compliqu√©es.</p>
<p>Certains packages peuvent grandement faciliter notre t√¢che lorsque
nous manipulons des donn√©es. Les packages dans R sont essentiellement
des ensembles de fonctions suppl√©mentaires qui vous permettent de faire
plus de choses. Les fonctions que nous avons utilis√©es jusqu‚Äô√† pr√©sent,
comme <code>str()</code> ou <code>data.frame()</code>, sont int√©gr√©es √†
R ; le chargement de packages donne acc√®s √† d‚Äôautres fonctions
sp√©cifiques. Avant d‚Äôutiliser un package pour la premi√®re fois, vous
devez l‚Äôinstaller sur votre machine, puis vous devez l‚Äôimporter √† chaque
session R suivante lorsque vous en avez besoin.</p>
<ul><li><p>Le package <strong><code>dplyr</code></strong> fournit des outils
puissants pour les t√¢ches de manipulation de donn√©es. Il est con√ßu pour
fonctionner directement avec des data frames, avec de nombreuses t√¢ches
de manipulation optimis√©es.</p></li>
<li><p>Comme nous le verrons plus loin, nous souhaitons parfois qu‚Äôun
data frame soit remodel√© pour pouvoir effectuer des analyses sp√©cifiques
ou pour la visualisation. Le package <strong><code>tidyr</code></strong>
r√©sout ce probl√®me courant de remodelage des donn√©es et fournit des
outils pour une manipulation <em>tidy</em> des donn√©es, c‚Äôest-√†-dire
selon les principes du <em>tidyverse</em>.</p></li>
</ul><p>Pour en savoir plus sur <strong><code>dplyr</code></strong> et
<strong><code>tidyr</code></strong>, vous voudrez consulter l‚Äô
aide-m√©moire sur la transformation de donn√©es pratique avec ** et celui
sur .</p>
<ul><li>Le package <strong><code>tidyverse</code></strong> est un ‚Äúpackage
parapluie‚Äù qui installe plusieurs packages utiles pour l‚Äôanalyse des
donn√©es qui fonctionnent bien ensemble, tels que
<strong><code>tidyr</code></strong>, * *<code>dplyr</code>**,
<strong><code>ggplot2</code></strong>,
<strong><code>tibble</code></strong>, etc. Ces packages nous aident √†
travailler et √† interagir avec les donn√©es. Ils nous permettent de faire
beaucoup de choses avec nos donn√©es, comme en extraire des
sous-ensembles, les transformer, les visualiser, etc.</li>
</ul><p>Si vous avez configur√© RStudio, vous devriez d√©j√† avoir install√© le
package tidyverse. V√©rifiez si vous l‚Äôavez en essayant de le charger
depuis la librarie (library)¬†:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## load the tidyverse packages, incl. dplyr</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></span></code></pre>
</div>
<p>Si vous recevez un message d‚Äôerreur
<code>aucun package nomm√© 'tidyverse' n'est trouv√©</code> alors vous
n‚Äôavez pas install√© le package pour cette version de R. Pour installer
le package <strong><code>tidyverse</code></strong>, tapez¬†:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></span></code></pre>
</div>
<p>Si vous avez d√ª installer le package
<strong><code>tidyverse</code></strong>, n‚Äôoubliez pas de le charger
dans cette session R en utilisant la commande <code>library()</code>
ci-dessus !</p>
</section><section><h2 class="section-heading" id="chargement-de-donn√©es-avec-le-tidyverse">Chargement de donn√©es avec le tidyverse<a class="anchor" aria-label="anchor" href="#chargement-de-donn%C3%A9es-avec-le-tidyverse"></a></h2>
<hr class="half-width"><p>Au lieu d‚Äôutiliser la fonction <code>read.csv()</code>, nous allons
lire nos donn√©es avec la fonction <code>read_csv()</code> (notez le
<code>_</code> au lieu du <code>.</code>) du package
<strong><code>readr</code></strong> du tidyverse.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/rnaseq.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## view the data</span></span>
<span><span class="va">rna</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 √ó 19
   gene    sample  expression organism   age sex   infection strain  time tissue
   &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 Asl     GSM254‚Ä¶       1170 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 2 Apod    GSM254‚Ä¶      36194 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 3 Cyp2d22 GSM254‚Ä¶       4060 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 4 Klk6    GSM254‚Ä¶        287 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 5 Fcrls   GSM254‚Ä¶         85 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 6 Slc2a4  GSM254‚Ä¶        782 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 7 Exd2    GSM254‚Ä¶       1619 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 8 Gjc2    GSM254‚Ä¶        288 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 9 Plp1    GSM254‚Ä¶      43217 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
10 Gnb4    GSM254‚Ä¶       1071 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
# ‚Ñπ 32,418 more rows
# ‚Ñπ 9 more variables: mouse &lt;dbl&gt;, ENTREZID &lt;dbl&gt;, product &lt;chr&gt;,
#   ensembl_gene_id &lt;chr&gt;, external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;,
#   gene_biotype &lt;chr&gt;, phenotype_description &lt;chr&gt;,
#   hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>Notez que la classe de la variable est d√©sormais appel√©e
‚Äútibble‚Äù.</p>
<p>Les tibbles modifient certains comportements des objets data frame
que nous avons introduits pr√©c√©demment. La structure des donn√©es est
tr√®s similaire √† un data frame. Pour nos besoins, les seules diff√©rences
sont les suivantes¬†:</p>
<ol style="list-style-type: decimal"><li><p>Il affiche le type de donn√©es de chaque colonne sous son nom.
Notez que &lt;<code>dbl</code>&gt; est un type de donn√©es d√©fini pour
contenir des valeurs num√©riques d√©cimales.</p></li>
<li><p>Il imprime uniquement les premi√®res lignes de donn√©es et
seulement les colonnes affichables √† l‚Äô√©cran.</p></li>
</ol><p>Nous allons maintenant apprendre les fonctions
<strong><code>dplyr</code></strong> les plus courantes¬†:</p>
<ul><li>
<code>select()</code>¬†: extraire des sous-ensembles de colonnes</li>
<li>
<code>filter()</code>¬†: extraire des sous-ensembles de lignes</li>
<li>
<code>mutate()</code>¬†: cr√©e de nouvelles colonnes en utilisant les
informations de colonnes pr√©existantes</li>
<li>
<code>group_by()</code> et <code>summarise()</code>¬†: cr√©ent des
statistiques r√©capitulatives sur des donn√©es group√©es</li>
<li>
<code>arrange()</code>¬†: trier les r√©sultats</li>
<li>
<code>count()</code>¬†: compte des valeurs discr√®tes</li>
</ul></section><section><h2 class="section-heading" id="s√©lection-de-colonnes-et-filtrage-de-lignes">S√©lection de colonnes et filtrage de lignes<a class="anchor" aria-label="anchor" href="#s%C3%A9lection-de-colonnes-et-filtrage-de-lignes"></a></h2>
<hr class="half-width"><p>Pour s√©lectionner les colonnes d‚Äôun bloc de donn√©es, utilisez
<code>select()</code>. Le premier argument de cette fonction est le
dataframe (<code>rna</code>), et les arguments suivants reprennent les
colonnes √† conserver.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span><span class="va">rna</span>, <span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 √ó 4
   gene    sample     tissue     expression
   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;
 1 Asl     GSM2545336 Cerebellum       1170
 2 Apod    GSM2545336 Cerebellum      36194
 3 Cyp2d22 GSM2545336 Cerebellum       4060
 4 Klk6    GSM2545336 Cerebellum        287
 5 Fcrls   GSM2545336 Cerebellum         85
 6 Slc2a4  GSM2545336 Cerebellum        782
 7 Exd2    GSM2545336 Cerebellum       1619
 8 Gjc2    GSM2545336 Cerebellum        288
 9 Plp1    GSM2545336 Cerebellum      43217
10 Gnb4    GSM2545336 Cerebellum       1071
# ‚Ñπ 32,418 more rows</code></pre>
</div>
<p>Pour s√©lectionner toutes les colonnes <em>sauf</em> certaines, mettez
un ‚Äú-‚Äù devant la variable √† exclure.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span><span class="va">rna</span>, <span class="op">-</span><span class="va">tissue</span>, <span class="op">-</span><span class="va">organism</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 √ó 17
   gene    sample   expression   age sex   infection strain  time mouse ENTREZID
   &lt;chr&gt;   &lt;chr&gt;         &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;
 1 Asl     GSM2545‚Ä¶       1170     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8    14   109900
 2 Apod    GSM2545‚Ä¶      36194     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8    14    11815
 3 Cyp2d22 GSM2545‚Ä¶       4060     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8    14    56448
 4 Klk6    GSM2545‚Ä¶        287     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8    14    19144
 5 Fcrls   GSM2545‚Ä¶         85     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8    14    80891
 6 Slc2a4  GSM2545‚Ä¶        782     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8    14    20528
 7 Exd2    GSM2545‚Ä¶       1619     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8    14    97827
 8 Gjc2    GSM2545‚Ä¶        288     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8    14   118454
 9 Plp1    GSM2545‚Ä¶      43217     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8    14    18823
10 Gnb4    GSM2545‚Ä¶       1071     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8    14    14696
# ‚Ñπ 32,418 more rows
# ‚Ñπ 7 more variables: product &lt;chr&gt;, ensembl_gene_id &lt;chr&gt;,
#   external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;, gene_biotype &lt;chr&gt;,
#   phenotype_description &lt;chr&gt;, hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>Cela s√©lectionnera toutes les variables de <code>rna</code> sauf
<code>tissue</code> et <code>organism</code>.</p>
<p>Pour choisir des lignes en fonction d‚Äôun crit√®re sp√©cifique, utilisez
<code>filter()</code>¬†:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">rna</span>, <span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 14,740 √ó 19
   gene    sample  expression organism   age sex   infection strain  time tissue
   &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 Asl     GSM254‚Ä¶        626 Mus mus‚Ä¶     8 Male  Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
 2 Apod    GSM254‚Ä¶      13021 Mus mus‚Ä¶     8 Male  Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
 3 Cyp2d22 GSM254‚Ä¶       2171 Mus mus‚Ä¶     8 Male  Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
 4 Klk6    GSM254‚Ä¶        448 Mus mus‚Ä¶     8 Male  Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
 5 Fcrls   GSM254‚Ä¶        180 Mus mus‚Ä¶     8 Male  Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
 6 Slc2a4  GSM254‚Ä¶        313 Mus mus‚Ä¶     8 Male  Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
 7 Exd2    GSM254‚Ä¶       2366 Mus mus‚Ä¶     8 Male  Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
 8 Gjc2    GSM254‚Ä¶        310 Mus mus‚Ä¶     8 Male  Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
 9 Plp1    GSM254‚Ä¶      53126 Mus mus‚Ä¶     8 Male  Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
10 Gnb4    GSM254‚Ä¶       1355 Mus mus‚Ä¶     8 Male  Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
# ‚Ñπ 14,730 more rows
# ‚Ñπ 9 more variables: mouse &lt;dbl&gt;, ENTREZID &lt;dbl&gt;, product &lt;chr&gt;,
#   ensembl_gene_id &lt;chr&gt;, external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;,
#   gene_biotype &lt;chr&gt;, phenotype_description &lt;chr&gt;,
#   hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">rna</span>, <span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span> <span class="op">&amp;</span> <span class="va">infection</span> <span class="op">==</span> <span class="st">"NonInfected"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 4,422 √ó 19
   gene    sample  expression organism   age sex   infection strain  time tissue
   &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 Asl     GSM254‚Ä¶        535 Mus mus‚Ä¶     8 Male  NonInfec‚Ä¶ C57BL‚Ä¶     0 Cereb‚Ä¶
 2 Apod    GSM254‚Ä¶      13668 Mus mus‚Ä¶     8 Male  NonInfec‚Ä¶ C57BL‚Ä¶     0 Cereb‚Ä¶
 3 Cyp2d22 GSM254‚Ä¶       2008 Mus mus‚Ä¶     8 Male  NonInfec‚Ä¶ C57BL‚Ä¶     0 Cereb‚Ä¶
 4 Klk6    GSM254‚Ä¶       1101 Mus mus‚Ä¶     8 Male  NonInfec‚Ä¶ C57BL‚Ä¶     0 Cereb‚Ä¶
 5 Fcrls   GSM254‚Ä¶        375 Mus mus‚Ä¶     8 Male  NonInfec‚Ä¶ C57BL‚Ä¶     0 Cereb‚Ä¶
 6 Slc2a4  GSM254‚Ä¶        249 Mus mus‚Ä¶     8 Male  NonInfec‚Ä¶ C57BL‚Ä¶     0 Cereb‚Ä¶
 7 Exd2    GSM254‚Ä¶       3126 Mus mus‚Ä¶     8 Male  NonInfec‚Ä¶ C57BL‚Ä¶     0 Cereb‚Ä¶
 8 Gjc2    GSM254‚Ä¶        791 Mus mus‚Ä¶     8 Male  NonInfec‚Ä¶ C57BL‚Ä¶     0 Cereb‚Ä¶
 9 Plp1    GSM254‚Ä¶      98658 Mus mus‚Ä¶     8 Male  NonInfec‚Ä¶ C57BL‚Ä¶     0 Cereb‚Ä¶
10 Gnb4    GSM254‚Ä¶       2437 Mus mus‚Ä¶     8 Male  NonInfec‚Ä¶ C57BL‚Ä¶     0 Cereb‚Ä¶
# ‚Ñπ 4,412 more rows
# ‚Ñπ 9 more variables: mouse &lt;dbl&gt;, ENTREZID &lt;dbl&gt;, product &lt;chr&gt;,
#   ensembl_gene_id &lt;chr&gt;, external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;,
#   gene_biotype &lt;chr&gt;, phenotype_description &lt;chr&gt;,
#   hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>Imaginons maintenant que nous nous int√©ressions aux homologues
humains des g√®nes de souris analys√©s dans ces donn√©es. Ces informations
se trouvent dans la derni√®re colonne du tibble <code>rna</code>, nomm√©e
<code>hsapiens_homolog_associated_gene_name</code>. Pour le visualiser
facilement, nous allons cr√©er un nouveau tableau contenant uniquement
les 2 colonnes <code>gene</code> et
<code>hsapiens_homolog_associated_gene_name</code>.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">rna</span>, <span class="va">gene</span>, <span class="va">hsapiens_homolog_associated_gene_name</span><span class="op">)</span></span>
<span><span class="va">genes</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 √ó 2
   gene    hsapiens_homolog_associated_gene_name
   &lt;chr&gt;   &lt;chr&gt;
 1 Asl     ASL
 2 Apod    APOD
 3 Cyp2d22 CYP2D6
 4 Klk6    KLK6
 5 Fcrls   FCRL2
 6 Slc2a4  SLC2A4
 7 Exd2    EXD2
 8 Gjc2    GJC2
 9 Plp1    PLP1
10 Gnb4    GNB4
# ‚Ñπ 32,418 more rows</code></pre>
</div>
<p>Certains g√®nes de souris n‚Äôont pas d‚Äôhomologues humains. Ceux-ci
peuvent √™tre r√©cup√©r√©s en utilisant <code>filter()</code> et la fonction
<code>is.na()</code>, qui d√©termine si quelque chose est un
<code>NA</code>.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">genes</span>, <span class="fu">is.na</span><span class="op">(</span><span class="va">hsapiens_homolog_associated_gene_name</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 4,290 √ó 2
   gene     hsapiens_homolog_associated_gene_name
   &lt;chr&gt;    &lt;chr&gt;
 1 Prodh    &lt;NA&gt;
 2 Tssk5    &lt;NA&gt;
 3 Vmn2r1   &lt;NA&gt;
 4 Gm10654  &lt;NA&gt;
 5 Hexa     &lt;NA&gt;
 6 Sult1a1  &lt;NA&gt;
 7 Gm6277   &lt;NA&gt;
 8 Tmem198b &lt;NA&gt;
 9 Adam1a   &lt;NA&gt;
10 Ebp      &lt;NA&gt;
# ‚Ñπ 4,280 more rows</code></pre>
</div>
<p>Si on veut conserver uniquement les g√®nes de souris qui ont un
homologue humain, on peut utiliser ‚Äú!‚Äù qui inverse le r√©sultat logique
pour demander chaque ligne o√π
<code>hsapiens\_homolog\_associated\_gene\_name</code> <em>n‚Äôest
pas</em> un <code>NA</code>.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">genes</span>, <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">hsapiens_homolog_associated_gene_name</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 28,138 √ó 2
   gene    hsapiens_homolog_associated_gene_name
   &lt;chr&gt;   &lt;chr&gt;
 1 Asl     ASL
 2 Apod    APOD
 3 Cyp2d22 CYP2D6
 4 Klk6    KLK6
 5 Fcrls   FCRL2
 6 Slc2a4  SLC2A4
 7 Exd2    EXD2
 8 Gjc2    GJC2
 9 Plp1    PLP1
10 Gnb4    GNB4
# ‚Ñπ 28,128 more rows</code></pre>
</div>
</section><section><h2 class="section-heading" id="pipes">Pipes<a class="anchor" aria-label="anchor" href="#pipes"></a></h2>
<hr class="half-width"><p>Comment faire si vous souhaitez s√©lectionner et filtrer en m√™me
temps¬†? Il existe trois fa√ßons de proc√©der¬†: utiliser des √©tapes
interm√©diaires, des fonctions imbriqu√©es ou des <em>pipes</em>.</p>
<p>Avec des √©tapes interm√©diaires, vous cr√©ez un data frame temporaire
et l‚Äôutilisez comme entr√©e de la fonction suivante, comme ceci¬†:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna2</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">rna</span>, <span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span></span>
<span><span class="va">rna3</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">rna2</span>, <span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span>
<span><span class="va">rna3</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 14,740 √ó 4
   gene    sample     tissue     expression
   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;
 1 Asl     GSM2545340 Cerebellum        626
 2 Apod    GSM2545340 Cerebellum      13021
 3 Cyp2d22 GSM2545340 Cerebellum       2171
 4 Klk6    GSM2545340 Cerebellum        448
 5 Fcrls   GSM2545340 Cerebellum        180
 6 Slc2a4  GSM2545340 Cerebellum        313
 7 Exd2    GSM2545340 Cerebellum       2366
 8 Gjc2    GSM2545340 Cerebellum        310
 9 Plp1    GSM2545340 Cerebellum      53126
10 Gnb4    GSM2545340 Cerebellum       1355
# ‚Ñπ 14,730 more rows</code></pre>
</div>
<p>Ceci est lisible, mais peut encombrer votre espace de travail avec de
nombreux objets interm√©diaires que vous devez nommer individuellement.
Avec beaucoup d‚Äô√©tapes, cette approche devient difficile √† suivre.</p>
<p>Vous pouvez √©galement imbriquer des fonctions (c‚Äôest-√†-dire une
fonction dans une autre), comme ceci¬†:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna3</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="fu">filter</span><span class="op">(</span><span class="va">rna</span>, <span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span>, <span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span>
<span><span class="va">rna3</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 14,740 √ó 4
   gene    sample     tissue     expression
   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;
 1 Asl     GSM2545340 Cerebellum        626
 2 Apod    GSM2545340 Cerebellum      13021
 3 Cyp2d22 GSM2545340 Cerebellum       2171
 4 Klk6    GSM2545340 Cerebellum        448
 5 Fcrls   GSM2545340 Cerebellum        180
 6 Slc2a4  GSM2545340 Cerebellum        313
 7 Exd2    GSM2545340 Cerebellum       2366
 8 Gjc2    GSM2545340 Cerebellum        310
 9 Plp1    GSM2545340 Cerebellum      53126
10 Gnb4    GSM2545340 Cerebellum       1355
# ‚Ñπ 14,730 more rows</code></pre>
</div>
<p>Cette approche est pratique, mais peut √™tre difficile √† lire si trop
de fonctions sont imbriqu√©es, car R √©value l‚Äôexpression de l‚Äôint√©rieur
vers l‚Äôext√©rieur (dans ce cas, filtrer, puis s√©lectionner).</p>
<p>La derni√®re option, le <em>pipe</em>, est un ajout plus r√©cent √† R.
Un <em>pipe</em> permet de prendre la sortie d‚Äôune fonction et de la
passer directement √† la suivante. Cette approche est utile lorsque vous
devez faire beaucoup d‚Äôop√©rations sur le m√™me jeu de donn√©es.</p>
<p>Les <em>pipes</em> dans R s‚Äô√©crivent √† <code>%&gt;%</code> (mis √†
disposition via le package <code>magrittr</code>) ou <code>|&gt;</code>
(via R). Si vous utilisez RStudio, vous pouvez afficher un <em>pipe</em>
avec <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd> sur un PC ou
<kbd>Cmd</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd> sur un Mac.</p>
<p>Dans le code ci-dessus, nous utilisons le <em>pipe</em> pour passer
les donn√©es <code>rna</code> d‚Äôabord dans <code>filter()</code> pour
conserver les lignes o√π <code>sex</code> est √©gal √† <code>"Male"</code>,
puis dans <code>select()</code> pour conserver uniquement les colonnes
<code>gene</code>, <code>sample</code>, <code>tissue</code> et
<code>expression</code>.</p>
<p>Le <em>pipe</em> <code>%&gt;%</code> prend l‚Äôobjet √† sa gauche et le
passe directement en tant que premier argument √† la fonction √† sa
droite. Nous n‚Äôavons donc plus besoin d‚Äôinclure explicitement les
donn√©es comme un argument pour les fonctions <code>filter()</code> et
<code>select()</code>.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 14,740 √ó 4
   gene    sample     tissue     expression
   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;
 1 Asl     GSM2545340 Cerebellum        626
 2 Apod    GSM2545340 Cerebellum      13021
 3 Cyp2d22 GSM2545340 Cerebellum       2171
 4 Klk6    GSM2545340 Cerebellum        448
 5 Fcrls   GSM2545340 Cerebellum        180
 6 Slc2a4  GSM2545340 Cerebellum        313
 7 Exd2    GSM2545340 Cerebellum       2366
 8 Gjc2    GSM2545340 Cerebellum        310
 9 Plp1    GSM2545340 Cerebellum      53126
10 Gnb4    GSM2545340 Cerebellum       1355
# ‚Ñπ 14,730 more rows</code></pre>
</div>
<p>Certains trouveront peut-√™tre utile de lire le <em>pipe</em> comme le
mot ‚Äúensuite‚Äù. Par exemple, dans l‚Äôexemple ci-dessus, nous avons pris le
data frame <code>rna</code>, <em>ensuite</em> nous avons
<code>filtr√©</code> pour les lignes avec <code>sex == "Male"</code>,
<em>ensuite</em> nous avons s√©lectionn√© les colonnes <code>gene</code>
<code>sample</code>, <code>tissue</code>, et
<code>expression</code>.</p>
<p>Les fonctions de <strong><code>dplyr</code></strong> sont
relativement simples, mais en les combinant avec le <em>pipe</em>, nous
pouvons accomplir des manipulations plus complexes.</p>
<p>Si nous voulons cr√©er un nouvel objet avec cette version plus petite
des donn√©es, il suffit de lui attribuer un nouveau nom¬†:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna3</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rna3</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 14,740 √ó 4
   gene    sample     tissue     expression
   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;           &lt;dbl&gt;
 1 Asl     GSM2545340 Cerebellum        626
 2 Apod    GSM2545340 Cerebellum      13021
 3 Cyp2d22 GSM2545340 Cerebellum       2171
 4 Klk6    GSM2545340 Cerebellum        448
 5 Fcrls   GSM2545340 Cerebellum        180
 6 Slc2a4  GSM2545340 Cerebellum        313
 7 Exd2    GSM2545340 Cerebellum       2366
 8 Gjc2    GSM2545340 Cerebellum        310
 9 Plp1    GSM2545340 Cerebellum      53126
10 Gnb4    GSM2545340 Cerebellum       1355
# ‚Ñπ 14,730 more rows</code></pre>
</div>
<div id="d√©fi" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="d√©fi" class="callout-inner">
<h3 class="callout-title">D√©fi:</h3>
<div class="callout-content">
<p>√Ä l‚Äôaide de <em>pipes</em>, extraire un sous-ensemble de
<code>rna</code> pour ne conserver que les observations des souris
femelles au temps 0, o√π le g√®ne a une expression sup√©rieure √† 50000, et
ne conservez que les colonnes <code>gene</code>, <code>sample</code>,
<code>time</code>, <code>expression</code> et <code>age</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">50000</span>,</span>
<span>         <span class="va">sex</span> <span class="op">==</span> <span class="st">"Female"</span>,</span>
<span>         <span class="va">time</span> <span class="op">==</span> <span class="fl">0</span> <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">time</span>, <span class="va">expression</span>, <span class="va">age</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 9 √ó 5
  gene   sample      time expression   age
  &lt;chr&gt;  &lt;chr&gt;      &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt;
1 Plp1   GSM2545337     0     101241     8
2 Atp1b1 GSM2545337     0      53260     8
3 Plp1   GSM2545338     0      96534     8
4 Atp1b1 GSM2545338     0      50614     8
5 Plp1   GSM2545348     0     102790     8
6 Atp1b1 GSM2545348     0      59544     8
7 Plp1   GSM2545353     0      71237     8
8 Glul   GSM2545353     0      52451     8
9 Atp1b1 GSM2545353     0      61451     8</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="mutate">Mutate<a class="anchor" aria-label="anchor" href="#mutate"></a></h2>
<hr class="half-width"><p>Vous souhaiterez fr√©quemment cr√©er de nouvelles colonnes bas√©es sur
les valeurs des colonnes existantes, par exemple pour effectuer des
conversions d‚Äôunit√©s ou pour trouver le rapport des valeurs dans deux
colonnes . Pour cela, nous utiliserons la fonction
<code>mutate()</code>.</p>
<p>Pour cr√©er une nouvelle colonne de temps en heures¬†:</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>time_hours <span class="op">=</span> <span class="va">time</span> <span class="op">*</span> <span class="fl">24</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">time</span>, <span class="va">time_hours</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 √ó 2
    time time_hours
   &lt;dbl&gt;      &lt;dbl&gt;
 1     8        192
 2     8        192
 3     8        192
 4     8        192
 5     8        192
 6     8        192
 7     8        192
 8     8        192
 9     8        192
10     8        192
# ‚Ñπ 32,418 more rows</code></pre>
</div>
<p>Vous pouvez √©galement cr√©er une deuxi√®me nouvelle colonne bas√©e sur
la premi√®re nouvelle colonne dans le m√™me appel de
<code>mutate()</code>¬†:</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>time_hours <span class="op">=</span> <span class="va">time</span> <span class="op">*</span> <span class="fl">24</span>,</span>
<span>         time_mn <span class="op">=</span> <span class="va">time_hours</span> <span class="op">*</span> <span class="fl">60</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">time</span>, <span class="va">time_hours</span>, <span class="va">time_mn</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 √ó 3
    time time_hours time_mn
   &lt;dbl&gt;      &lt;dbl&gt;   &lt;dbl&gt;
 1     8        192   11520
 2     8        192   11520
 3     8        192   11520
 4     8        192   11520
 5     8        192   11520
 6     8        192   11520
 7     8        192   11520
 8     8        192   11520
 9     8        192   11520
10     8        192   11520
# ‚Ñπ 32,418 more rows</code></pre>
</div>
<div id="d√©fi-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="d√©fi-1" class="callout-inner">
<h3 class="callout-title">D√©fi</h3>
<div class="callout-content">
<p>Cr√©ez un nouveau data frane √† partir de <code>rna</code> qui r√©pond
aux crit√®res suivants¬†: contient uniquement les colonnes
<code>gene</code>, <code>chromosome_name</code>,
<code>phenotype_description</code>, <code>sample</code> et
<code>expression</code>. Les valeurs d‚Äôexpression doivent √™tre
transform√©es en log. Ce data frame doit contenir uniquement des g√®nes
situ√©s sur les chromosomes sexuels, ayant bien un
<code>phenotype \_description</code>, et une expression en log
sup√©rieure √† 5.</p>
<p><strong>Astuce</strong>¬†: r√©fl√©chissez bien √† la fa√ßon dont les
commandes doivent √™tre ordonn√©es !</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>expression <span class="op">=</span> <span class="fu">log</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">chromosome_name</span>, <span class="va">phenotype_description</span>, <span class="va">sample</span>, <span class="va">expression</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"X"</span> <span class="op">|</span> <span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"Y"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">phenotype_description</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 649 √ó 5
   gene   chromosome_name phenotype_description                sample expression
   &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;                                &lt;chr&gt;       &lt;dbl&gt;
 1 Plp1   X               abnormal CNS glial cell morphology   GSM25‚Ä¶      10.7
 2 Slc7a3 X               decreased body length                GSM25‚Ä¶       5.46
 3 Plxnb3 X               abnormal coat appearance             GSM25‚Ä¶       6.58
 4 Rbm3   X               abnormal liver morphology            GSM25‚Ä¶       9.32
 5 Cfp    X               abnormal cardiovascular system phys‚Ä¶ GSM25‚Ä¶       6.18
 6 Ebp    X               abnormal embryonic erythrocyte morp‚Ä¶ GSM25‚Ä¶       6.68
 7 Cd99l2 X               abnormal cellular extravasation      GSM25‚Ä¶       8.04
 8 Piga   X               abnormal brain development           GSM25‚Ä¶       6.06
 9 Pim2   X               decreased T cell proliferation       GSM25‚Ä¶       7.11
10 Itm2a  X               no abnormal phenotype detected       GSM25‚Ä¶       7.48
# ‚Ñπ 639 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="analyse-de-donn√©es-split-apply-combine">Analyse de donn√©es ‚Äúsplit-apply-combine‚Äù<a class="anchor" aria-label="anchor" href="#analyse-de-donn%C3%A9es-split-apply-combine"></a></h2>
<hr class="half-width"><p>De nombreuses t√¢ches d‚Äôanalyse de donn√©es peuvent √™tre abord√©es √†
l‚Äôaide du paradigme <em>split-apply-combine</em>¬†: divisez les donn√©es
en groupes, appliquez une analyse √† chaque groupe, puis combinez les
r√©sultats. <strong><code>dplyr</code></strong> rend cela tr√®s facile
gr√¢ce √† l‚Äôutilisation de la fonction <code>group_by()</code>.</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 √ó 19
# Groups:   gene [1,474]
   gene    sample  expression organism   age sex   infection strain  time tissue
   &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 Asl     GSM254‚Ä¶       1170 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 2 Apod    GSM254‚Ä¶      36194 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 3 Cyp2d22 GSM254‚Ä¶       4060 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 4 Klk6    GSM254‚Ä¶        287 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 5 Fcrls   GSM254‚Ä¶         85 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 6 Slc2a4  GSM254‚Ä¶        782 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 7 Exd2    GSM254‚Ä¶       1619 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 8 Gjc2    GSM254‚Ä¶        288 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 9 Plp1    GSM254‚Ä¶      43217 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
10 Gnb4    GSM254‚Ä¶       1071 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
# ‚Ñπ 32,418 more rows
# ‚Ñπ 9 more variables: mouse &lt;dbl&gt;, ENTREZID &lt;dbl&gt;, product &lt;chr&gt;,
#   ensembl_gene_id &lt;chr&gt;, external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;,
#   gene_biotype &lt;chr&gt;, phenotype_description &lt;chr&gt;,
#   hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>La fonction <code>group_by()</code> n‚Äôeffectue aucun traitement de
donn√©es, elle regroupe simplement les donn√©es en sous-ensembles : dans
l‚Äôexemple ci-dessus, notre <code>tibble</code> initial de 32428
observations est divis√© en 1474 groupes selon la variable
<code>gene</code>.</p>
<p>On pourrait de m√™me d√©cider de regrouper le tibble par
√©chantillons¬†:</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 √ó 19
# Groups:   sample [22]
   gene    sample  expression organism   age sex   infection strain  time tissue
   &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 Asl     GSM254‚Ä¶       1170 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 2 Apod    GSM254‚Ä¶      36194 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 3 Cyp2d22 GSM254‚Ä¶       4060 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 4 Klk6    GSM254‚Ä¶        287 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 5 Fcrls   GSM254‚Ä¶         85 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 6 Slc2a4  GSM254‚Ä¶        782 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 7 Exd2    GSM254‚Ä¶       1619 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 8 Gjc2    GSM254‚Ä¶        288 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 9 Plp1    GSM254‚Ä¶      43217 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
10 Gnb4    GSM254‚Ä¶       1071 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
# ‚Ñπ 32,418 more rows
# ‚Ñπ 9 more variables: mouse &lt;dbl&gt;, ENTREZID &lt;dbl&gt;, product &lt;chr&gt;,
#   ensembl_gene_id &lt;chr&gt;, external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;,
#   gene_biotype &lt;chr&gt;, phenotype_description &lt;chr&gt;,
#   hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>Ici, notre <code>tibble</code> initial de 32428 observations est
divis√© en 22 groupes selon la variable <code>sample</code>.</p>
<p>Une fois les donn√©es regroup√©es, les op√©rations qui suivent sont
appliqu√©es sur chaque groupe ind√©pendamment.</p>
<div class="section level3">
<h3 id="la-fonction-summarise">La fonction <code>summarise()</code>
<a class="anchor" aria-label="anchor" href="#la-fonction-summarise"></a></h3>
<p><code>group_by()</code> est souvent utilis√© avec
<code>summarise()</code>, qui r√©duit chaque groupe en un r√©sum√© d‚Äôune
seule ligne.</p>
<p><code>group_by()</code> prend comme arguments les noms de colonnes
qui contiennent les variables <strong>cat√©gorielles</strong> pour
lesquelles vous souhaitez calculer les statistiques r√©capitulatives.
Donc, pour calculer l‚Äôexpression moyenne par g√®ne¬†:</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 √ó 2
   gene     mean_expression
   &lt;chr&gt;              &lt;dbl&gt;
 1 AI504432         1053.
 2 AW046200          131.
 3 AW551984          295.
 4 Aamp             4751.
 5 Abca12              4.55
 6 Abcc8            2498.
 7 Abhd14a           525.
 8 Abi2             4909.
 9 Abi3bp           1002.
10 Abl2             2124.
# ‚Ñπ 1,464 more rows</code></pre>
</div>
<p>Nous pourrions √©galement vouloir calculer les niveaux d‚Äôexpression
moyens de tous les g√®nes dans chaque √©chantillon¬†:</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 22 √ó 2
   sample     mean_expression
   &lt;chr&gt;                &lt;dbl&gt;
 1 GSM2545336           2062.
 2 GSM2545337           1766.
 3 GSM2545338           1668.
 4 GSM2545339           1696.
 5 GSM2545340           1682.
 6 GSM2545341           1638.
 7 GSM2545342           1594.
 8 GSM2545343           2107.
 9 GSM2545344           1712.
10 GSM2545345           1700.
# ‚Ñπ 12 more rows</code></pre>
</div>
<p>On peut √©galement regrouper les observations par plusieurs
colonnes¬†:</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene', 'infection'. You can override using
the `.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 4,422 √ó 4
# Groups:   gene, infection [2,948]
   gene     infection    time mean_expression
   &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt;           &lt;dbl&gt;
 1 AI504432 InfluenzaA      4           1104.
 2 AI504432 InfluenzaA      8           1014
 3 AI504432 NonInfected     0           1034.
 4 AW046200 InfluenzaA      4            152.
 5 AW046200 InfluenzaA      8             81
 6 AW046200 NonInfected     0            155.
 7 AW551984 InfluenzaA      4            302.
 8 AW551984 InfluenzaA      8            342.
 9 AW551984 NonInfected     0            238
10 Aamp     InfluenzaA      4           4870
# ‚Ñπ 4,412 more rows</code></pre>
</div>
<p>Une fois les donn√©es regroup√©es, vous pouvez √©galement r√©sumer
plusieurs variables en m√™me temps (et pas n√©cessairement sur la m√™me
variable). Par exemple, nous pourrions ajouter une colonne indiquant
l‚Äôexpression m√©diane par g√®ne et par condition¬†:</p>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span>,</span>
<span>            median_expression <span class="op">=</span> <span class="fu">median</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene', 'infection'. You can override using
the `.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 4,422 √ó 5
# Groups:   gene, infection [2,948]
   gene     infection    time mean_expression median_expression
   &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt;           &lt;dbl&gt;             &lt;dbl&gt;
 1 AI504432 InfluenzaA      4           1104.             1094.
 2 AI504432 InfluenzaA      8           1014               985
 3 AI504432 NonInfected     0           1034.             1016
 4 AW046200 InfluenzaA      4            152.              144.
 5 AW046200 InfluenzaA      8             81                82
 6 AW046200 NonInfected     0            155.              163
 7 AW551984 InfluenzaA      4            302.              245
 8 AW551984 InfluenzaA      8            342.              287
 9 AW551984 NonInfected     0            238               265
10 Aamp     InfluenzaA      4           4870              4708
# ‚Ñπ 4,412 more rows</code></pre>
</div>
<div id="d√©fi-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="d√©fi-2" class="callout-inner">
<h3 class="callout-title">D√©fi</h3>
<div class="callout-content">
<p>Calculer le niveau d‚Äôexpression moyen du g√®ne ‚ÄúDok3‚Äù pour chaque
temps.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">gene</span> <span class="op">==</span> <span class="st">"Dok3"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 √ó 2
   time  mean
  &lt;dbl&gt; &lt;dbl&gt;
1     0  169
2     4  156.
3     8   61 </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="count">Count<a class="anchor" aria-label="anchor" href="#count"></a></h3>
<p>Lorsque nous travaillons avec des donn√©es, nous souhaitons souvent
conna√Ætre le nombre d‚Äôobservations trouv√©es pour chaque facteur ou
combinaison de facteurs. Pour cette t√¢che,
<strong><code>dplyr</code></strong> fournit la fonction
<code>count()</code>. Par exemple, si nous voulons compter le nombre
d‚Äôobservations pour chaque √©chantillon infect√© et non infect√©¬†:</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">count</span><span class="op">(</span><span class="va">infection</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 2 √ó 2
  infection       n
  &lt;chr&gt;       &lt;int&gt;
1 InfluenzaA  22110
2 NonInfected 10318</code></pre>
</div>
<p>La fonction <code>count()</code> est un raccourci pour quelque chose
que nous avons d√©j√† vu : regrouper par une variable et la r√©sumer en
comptant le nombre d‚Äôobservations dans ce groupe. En d‚Äôautres termes,
<code>rna %&gt;% count(infection)</code> √©quivaut √†¬†:</p>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">group_by</span><span class="op">(</span><span class="va">infection</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">summarise</span><span class="op">(</span>n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 2 √ó 2
  infection       n
  &lt;chr&gt;       &lt;int&gt;
1 InfluenzaA  22110
2 NonInfected 10318</code></pre>
</div>
<p>L‚Äôexemple pr√©c√©dent montre l‚Äôutilisation de <code>count()</code> pour
compter le nombre de lignes/observations pour <em>un</em> facteur
(c‚Äôest-√†-dire <code>infection</code>). Pour compter une <em>combinaison
de facteurs</em>, telle que <code>infection</code> et <code>time</code>,
il suffit de sp√©cifier les deux facteurs comme arguments de
<code>count()</code>¬†:</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">count</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 √ó 3
  infection    time     n
  &lt;chr&gt;       &lt;dbl&gt; &lt;int&gt;
1 InfluenzaA      4 11792
2 InfluenzaA      8 10318
3 NonInfected     0 10318</code></pre>
</div>
<p>ce qui √©quivaut √† ceci :</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'infection'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 √ó 3
# Groups:   infection [2]
  infection    time     n
  &lt;chr&gt;       &lt;dbl&gt; &lt;int&gt;
1 InfluenzaA      4 11792
2 InfluenzaA      8 10318
3 NonInfected     0 10318</code></pre>
</div>
<p>Il est parfois utile de trier le r√©sultat pour faciliter les
comparaisons. Nous pouvons utiliser <code>arrange()</code> pour trier le
tableau. Par exemple, nous pourrions vouloir organiser le tableau
ci-dessus par temps¬†:</p>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">time</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 √ó 3
  infection    time     n
  &lt;chr&gt;       &lt;dbl&gt; &lt;int&gt;
1 NonInfected     0 10318
2 InfluenzaA      4 11792
3 InfluenzaA      8 10318</code></pre>
</div>
<p>ou par comptage¬†:</p>
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">n</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 √ó 3
  infection    time     n
  &lt;chr&gt;       &lt;dbl&gt; &lt;int&gt;
1 InfluenzaA      8 10318
2 NonInfected     0 10318
3 InfluenzaA      4 11792</code></pre>
</div>
<p>Pour trier par ordre d√©croissant, nous devons ajouter la fonction
<code>desc()</code>¬†:</p>
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 √ó 3
  infection    time     n
  &lt;chr&gt;       &lt;dbl&gt; &lt;int&gt;
1 InfluenzaA      4 11792
2 InfluenzaA      8 10318
3 NonInfected     0 10318</code></pre>
</div>
<div id="d√©fi-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="d√©fi-3" class="callout-inner">
<h3 class="callout-title">D√©fi</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>Combien de g√®nes ont √©t√© analys√©s dans chaque √©chantillon¬†?</li>
<li>Utilisez <code>group_by()</code> et <code>summarise()</code> pour
√©valuer la profondeur de s√©quen√ßage (la somme de tous les comptes) dans
chaque √©chantillon. Quel √©chantillon a la profondeur de s√©quen√ßage la
plus √©lev√©e¬†?</li>
<li>Choisissez un √©chantillon et √©valuez le nombre de g√®nes par
biotype.</li>
<li>Identifiez les g√®nes associ√©s √† la description du ph√©notype
‚Äúabnormal DNA methylation‚Äù et calculez leur expression moyenne (en log)
au temps 0, 4 et 8.</li>
</ol></div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb66">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1.</span></span>
<span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 22 √ó 2
   sample         n
   &lt;chr&gt;      &lt;int&gt;
 1 GSM2545336  1474
 2 GSM2545337  1474
 3 GSM2545338  1474
 4 GSM2545339  1474
 5 GSM2545340  1474
 6 GSM2545341  1474
 7 GSM2545342  1474
 8 GSM2545343  1474
 9 GSM2545344  1474
10 GSM2545345  1474
# ‚Ñπ 12 more rows</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb68">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 2.</span></span>
<span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>seq_depth <span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">seq_depth</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 22 √ó 2
   sample     seq_depth
   &lt;chr&gt;          &lt;dbl&gt;
 1 GSM2545350   3255566
 2 GSM2545352   3216163
 3 GSM2545343   3105652
 4 GSM2545336   3039671
 5 GSM2545380   3036098
 6 GSM2545353   2953249
 7 GSM2545348   2913678
 8 GSM2545362   2913517
 9 GSM2545351   2782464
10 GSM2545349   2758006
# ‚Ñπ 12 more rows</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb70">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 3.</span></span>
<span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sample</span> <span class="op">==</span> <span class="st">"GSM2545336"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">gene_biotype</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 13 √ó 2
   gene_biotype                           n
   &lt;chr&gt;                              &lt;int&gt;
 1 protein_coding                      1321
 2 lncRNA                                69
 3 processed_pseudogene                  59
 4 miRNA                                  7
 5 snoRNA                                 5
 6 TEC                                    4
 7 polymorphic_pseudogene                 2
 8 unprocessed_pseudogene                 2
 9 IG_C_gene                              1
10 scaRNA                                 1
11 transcribed_processed_pseudogene       1
12 transcribed_unitary_pseudogene         1
13 transcribed_unprocessed_pseudogene     1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb72">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 4.</span></span>
<span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">phenotype_description</span> <span class="op">==</span> <span class="st">"abnormal DNA methylation"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="fu">log</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 √ó 3
# Groups:   gene [2]
  gene   time mean_expression
  &lt;chr&gt; &lt;dbl&gt;           &lt;dbl&gt;
1 Xist      0            6.95
2 Xist      4            6.34
3 Xist      8            7.13
4 Zdbf2     0            6.27
5 Zdbf2     4            6.27
6 Zdbf2     8            6.19</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="remodeler-les-donn√©es">Remodeler les donn√©es<a class="anchor" aria-label="anchor" href="#remodeler-les-donn%C3%A9es"></a></h2>
<hr class="half-width"><p>Dans le tibble <code>rna</code>, les lignes contiennent des valeurs
d‚Äôexpression qui sont associ√©es √† une combinaison de 2 autres variables
: <code>gene</code> et <code>sample</code>.</p>
<p>Toutes les autres colonnes correspondent √† des variables d√©crivant
soit l‚Äô√©chantillon (organisme, √¢ge, sexe, ‚Ä¶) soit le g√®ne
(‚Äògene_biotype‚Äô, ‚ÄòENTREZ_ID‚Äô, ‚Äòproduct‚Äô, ‚Ä¶). Les variables qui ne
changent pas avec les g√®nes ou avec les √©chantillons auront la m√™me
valeur dans toutes les lignes.</p>
<div class="codewrapper sourceCode" id="cb75">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 √ó 19
   gene     sample expression organism   age sex   infection strain  time tissue
   &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;
 1 AI504432 GSM25‚Ä¶       1230 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 2 AI504432 GSM25‚Ä¶       1085 Mus mus‚Ä¶     8 Fema‚Ä¶ NonInfec‚Ä¶ C57BL‚Ä¶     0 Cereb‚Ä¶
 3 AI504432 GSM25‚Ä¶        969 Mus mus‚Ä¶     8 Fema‚Ä¶ NonInfec‚Ä¶ C57BL‚Ä¶     0 Cereb‚Ä¶
 4 AI504432 GSM25‚Ä¶       1284 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
 5 AI504432 GSM25‚Ä¶        966 Mus mus‚Ä¶     8 Male  Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
 6 AI504432 GSM25‚Ä¶        918 Mus mus‚Ä¶     8 Male  Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 7 AI504432 GSM25‚Ä¶        985 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     8 Cereb‚Ä¶
 8 AI504432 GSM25‚Ä¶        972 Mus mus‚Ä¶     8 Male  NonInfec‚Ä¶ C57BL‚Ä¶     0 Cereb‚Ä¶
 9 AI504432 GSM25‚Ä¶       1000 Mus mus‚Ä¶     8 Fema‚Ä¶ Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
10 AI504432 GSM25‚Ä¶        816 Mus mus‚Ä¶     8 Male  Influenz‚Ä¶ C57BL‚Ä¶     4 Cereb‚Ä¶
# ‚Ñπ 32,418 more rows
# ‚Ñπ 9 more variables: mouse &lt;dbl&gt;, ENTREZID &lt;dbl&gt;, product &lt;chr&gt;,
#   ensembl_gene_id &lt;chr&gt;, external_synonym &lt;chr&gt;, chromosome_name &lt;chr&gt;,
#   gene_biotype &lt;chr&gt;, phenotype_description &lt;chr&gt;,
#   hsapiens_homolog_associated_gene_name &lt;chr&gt;</code></pre>
</div>
<p>Cette structure est appel√©e ‚Äúformat long‚Äù, car une colonne contient
toutes les valeurs, et d‚Äôautres colonnes r√©pertorient le contexte de la
valeur.</p>
<p>Dans certains cas, le ‚Äúformat long‚Äù n‚Äôest pas facile √† lire, et un
autre format, un ‚Äúformat large‚Äù est pr√©f√©r√©, comme mani√®re plus compacte
de repr√©senter les donn√©es. Cela est typiquement le cas pour des donn√©es
d‚Äôexpression de g√®nes, qui sont g√©n√©ralement repr√©sent√©es sous forme de
matrice o√π les g√®nes sont repris le long des lignes et les √©chantillons
le long des colonnes.</p>
<p>Dans ce format, il devient donc simple d‚Äôexplorer la relation entre
les niveaux d‚Äôexpression g√©nique au sein et entre les √©chantillons.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 √ó 23
   gene    GSM2545336 GSM2545337 GSM2545338 GSM2545339 GSM2545340 GSM2545341
   &lt;chr&gt;        &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
 1 Asl           1170        361        400        586        626        988
 2 Apod         36194      10347       9173      10620      13021      29594
 3 Cyp2d22       4060       1616       1603       1901       2171       3349
 4 Klk6           287        629        641        578        448        195
 5 Fcrls           85        233        244        237        180         38
 6 Slc2a4         782        231        248        265        313        786
 7 Exd2          1619       2288       2235       2513       2366       1359
 8 Gjc2           288        595        568        551        310        146
 9 Plp1         43217     101241      96534      58354      53126      27173
10 Gnb4          1071       1791       1867       1430       1355        798
# ‚Ñπ 1,464 more rows
# ‚Ñπ 16 more variables: GSM2545342 &lt;dbl&gt;, GSM2545343 &lt;dbl&gt;, GSM2545344 &lt;dbl&gt;,
#   GSM2545345 &lt;dbl&gt;, GSM2545346 &lt;dbl&gt;, GSM2545347 &lt;dbl&gt;, GSM2545348 &lt;dbl&gt;,
#   GSM2545349 &lt;dbl&gt;, GSM2545350 &lt;dbl&gt;, GSM2545351 &lt;dbl&gt;, GSM2545352 &lt;dbl&gt;,
#   GSM2545353 &lt;dbl&gt;, GSM2545354 &lt;dbl&gt;, GSM2545362 &lt;dbl&gt;, GSM2545363 &lt;dbl&gt;,
#   GSM2545380 &lt;dbl&gt;</code></pre>
</div>
<p>Pour convertir les valeurs d‚Äôexpression g√©nique de <code>rna</code>
en un format large, nous devons cr√©er une nouvelle table o√π les valeurs
de la colonne <code>sample</code> deviendraient les noms des variables
de colonne.</p>
<p>Nous <strong>remodellons</strong> les donn√©es en fonction des
observations d‚Äôint√©r√™t, tout en continant √† suivre les principles des
donn√©es tidy: niveaux d‚Äôexpression par g√®ne au lieu des les afficher par
g√®nes et par √©chantillons.</p>
<p>La transformation inverse serait de transformer les noms de colonnes
en valeurs d‚Äôune nouvelle variable.</p>
<p>Nous pouvons effectuer ces deux transformations avec deux fonctions
du package <code>tidyr</code>, √† savoir <code>pivot_longer()</code> et
<code>pivot_wider()</code> (voir <a href="https://tidyr.tidyverse.org%20/dev/articles/pivot.html" class="external-link">ici</a>
pour d√©tails).</p>
<div class="section level3">
<h3 id="pivoter-les-donn√©es-en-format-large">Pivoter les donn√©es en format large<a class="anchor" aria-label="anchor" href="#pivoter-les-donn%C3%A9es-en-format-large"></a></h3>
<p>S√©lectionnons les 3 premi√®res colonnes de <code>rna</code> et
utilisons <code>pivot_wider()</code> pour transformer les donn√©es en
format large.</p>
<div class="codewrapper sourceCode" id="cb78">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_exp</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">expression</span><span class="op">)</span></span>
<span><span class="va">rna_exp</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 √ó 3
   gene    sample     expression
   &lt;chr&gt;   &lt;chr&gt;           &lt;dbl&gt;
 1 Asl     GSM2545336       1170
 2 Apod    GSM2545336      36194
 3 Cyp2d22 GSM2545336       4060
 4 Klk6    GSM2545336        287
 5 Fcrls   GSM2545336         85
 6 Slc2a4  GSM2545336        782
 7 Exd2    GSM2545336       1619
 8 Gjc2    GSM2545336        288
 9 Plp1    GSM2545336      43217
10 Gnb4    GSM2545336       1071
# ‚Ñπ 32,418 more rows</code></pre>
</div>
<p><code>pivot_wider</code> prend trois arguments principaux¬†:</p>
<ol style="list-style-type: decimal"><li>les donn√©es √† transformer ;</li>
<li>le <code>names_from</code> : la colonne dont les valeurs deviendront
les nouveaux noms de colonne ;</li>
<li>les <code>values_from</code>¬†: la colonne dont les valeurs
rempliront les nouvelles colonnes.</li>
</ol><figure style="text-align: center"><img src="../fig/pivot_wider.png" alt="Grand pivot des donn√©es `rna`." class="figure mx-auto d-block"><figcaption>
Grand pivot des donn√©es <code>rna</code>.
</figcaption></figure><div class="codewrapper sourceCode" id="cb80">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_wide</span> <span class="op">&lt;-</span> <span class="va">rna_exp</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sample</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">expression</span><span class="op">)</span></span>
<span><span class="va">rna_wide</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 √ó 23
   gene    GSM2545336 GSM2545337 GSM2545338 GSM2545339 GSM2545340 GSM2545341
   &lt;chr&gt;        &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
 1 Asl           1170        361        400        586        626        988
 2 Apod         36194      10347       9173      10620      13021      29594
 3 Cyp2d22       4060       1616       1603       1901       2171       3349
 4 Klk6           287        629        641        578        448        195
 5 Fcrls           85        233        244        237        180         38
 6 Slc2a4         782        231        248        265        313        786
 7 Exd2          1619       2288       2235       2513       2366       1359
 8 Gjc2           288        595        568        551        310        146
 9 Plp1         43217     101241      96534      58354      53126      27173
10 Gnb4          1071       1791       1867       1430       1355        798
# ‚Ñπ 1,464 more rows
# ‚Ñπ 16 more variables: GSM2545342 &lt;dbl&gt;, GSM2545343 &lt;dbl&gt;, GSM2545344 &lt;dbl&gt;,
#   GSM2545345 &lt;dbl&gt;, GSM2545346 &lt;dbl&gt;, GSM2545347 &lt;dbl&gt;, GSM2545348 &lt;dbl&gt;,
#   GSM2545349 &lt;dbl&gt;, GSM2545350 &lt;dbl&gt;, GSM2545351 &lt;dbl&gt;, GSM2545352 &lt;dbl&gt;,
#   GSM2545353 &lt;dbl&gt;, GSM2545354 &lt;dbl&gt;, GSM2545362 &lt;dbl&gt;, GSM2545363 &lt;dbl&gt;,
#   GSM2545380 &lt;dbl&gt;</code></pre>
</div>
<p>Notez que par d√©faut, la fonction <code>pivot_wider()</code> ajoutera
<code>NA</code> pour les valeurs manquantes.</p>
<p>Imaginons que nous ayons des valeurs d‚Äôexpression manquantes pour
certains g√®nes dans certains √©chantillons. Dans cet exemple, le g√®ne
Cyp2d22 n‚Äôaune donn√©e d‚Äôexpression que pour l‚Äôechantillon
GSM2545338.</p>
<p>Par d√©faut, la fonction <code>pivot_wider()</code> ajoutera
<code>NA</code> pour les valeurs manquantes. Ceci peut √™tre param√©tr√©
avec l‚Äôargument <code>values_fill</code> de la fonction
<code>pivot_wider()</code>.</p>
<div class="codewrapper sourceCode" id="cb82">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_with_missing_values</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">√©chantillon</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'rna_with_missing_values' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb84">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_with_missing_values</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">√©chantillon</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">expression</span>,</span>
<span>              valeurs_fill <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'rna_with_missing_values' not found</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="pivoter-les-donn√©es-en-format-long">Pivoter les donn√©es en format long<a class="anchor" aria-label="anchor" href="#pivoter-les-donn%C3%A9es-en-format-long"></a></h3>
<p>Dans la situation inverse, nous utilisons les noms de colonnes et les
transformons en deux nouvelles variables. Une variable repr√©sente le nom
de colonne contenant les valeurs, et l‚Äôautre contenant les valeurs
pr√©c√©demment associ√©es aux noms de colonnes.</p>
<p><code>pivot_longer()</code> prend quatre arguments principaux¬†:</p>
<ol style="list-style-type: decimal"><li>les donn√©es √† transformer ;</li>
<li>
<code>names_to</code>¬†: le nouveau nom de colonne que nous
souhaitons cr√©er et remplir avec les noms de colonnes actuels¬†;</li>
<li>
<code>values_to</code>¬†: le nouveau nom de colonne que nous
souhaitons cr√©er et remplir avec valeurs actuelles¬†;</li>
<li>les noms des colonnes √† utiliser pour renseigner les variables
<code>names_to</code> et <code>values_to</code> (ou √† ignorer).</li>
</ol><p><code>{r, fig.cap="Pivot long des donn√©es `rna`.", echo=FALSE, message=FALSE} knitr::include_graphics("fig/pivot_longer.png")</code></p>
<pre><code>
Pour cr√©er `rna_long` √† partir de `rna_wide`, nous cr√©ons une premi√®re variable `"sample"` qui conteindra les noms des √©chantillons repris comme noms de colonnes (hormi `gene`), et une seconde, `"expression"`, qui contiendra les valeurs de la table. Nous ommettons la colonne `gene` pour la transformation en y ajoutant le signe moins.

Notez que les nouveaux noms de colonnes sont repris entre guillemets.


``` r
rna_long &lt;- rna_wide %&gt;%
    pivot_longer(names_to = "sample",
                 values_to = "expression",
                 -gene)
rna_long</code></pre>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 √ó 3
   gene  sample     expression
   &lt;chr&gt; &lt;chr&gt;           &lt;dbl&gt;
 1 Asl   GSM2545336       1170
 2 Asl   GSM2545337        361
 3 Asl   GSM2545338        400
 4 Asl   GSM2545339        586
 5 Asl   GSM2545340        626
 6 Asl   GSM2545341        988
 7 Asl   GSM2545342        836
 8 Asl   GSM2545343        535
 9 Asl   GSM2545344        586
10 Asl   GSM2545345        597
# ‚Ñπ 32,418 more rows</code></pre>
</div>
<p>Nous aurions √©galement pu sp√©cifier les colonnes √† inclure. Quand il
faut inclure un grand nombre de colonnes, et il est plus facile de
sp√©cifier ce qu‚Äôil faut rassembler que ce qu‚Äôil faut ignorer. Ici, la
fonction <code>starts_with()</code> peut aider √† r√©cup√©rer des colonnes
sans avoir √† toutes les √©num√©rer ! Une autre possibilit√© serait
d‚Äôutiliser l‚Äôop√©rateur <code>:</code> !</p>
<div class="codewrapper sourceCode" id="cb88">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_wide</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                 values_to <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                 cols <span class="op">=</span> <span class="fu">start_with</span><span class="op">(</span><span class="st">"GSM"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `pivot_longer()`:
‚Ñπ In argument: `start_with("GSM")`.
Caused by error in `start_with()`:
! could not find function "start_with"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb90">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_wide</span> <span class="op">%&gt; %</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                 valeurs_to <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                 <span class="va">GSM2545336</span><span class="op">:</span><span class="va">GSM2545380</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in rna_wide %&gt; % pivot_longer(names_to = "sample", valeurs_to = "expression", : could not find function "%&gt; %"</code></pre>
</div>
<p>Notez que si nous avions des valeurs manquantes dans le format large,
le <code>NA</code> serait inclus dans le nouveau format long.</p>
<p>Souvenez-vous de notre pr√©c√©dent tibble fictif contenant des valeurs
manquantes¬†:</p>
<div class="codewrapper sourceCode" id="cb92">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_with_missing_values</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'rna_with_missing_values' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb94">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wide_with_NA</span> <span class="op">&lt;-</span> <span class="va">rna_with_missing_values</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sample</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'rna_with_missing_values' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb96">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wide_with_NA</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'wide_with_NA' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb98">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wide_with_NA</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                 values_to <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                 <span class="op">-</span><span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'wide_with_NA' not found</code></pre>
</div>
<p>Passer √† des formats larges et longs peut √™tre un moyen utile
d‚Äô√©quilibrer un ensemble de donn√©es afin que chaque r√©plique ait la m√™me
composition.</p>
<div id="question" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="question" class="callout-inner">
<h3 class="callout-title">Question</h3>
<div class="callout-content">
<p>A partir de la table <code>rna</code>, utilisez la fonction
<code>pivot_wider()</code> pour cr√©er un tableau au format large donnant
les niveaux d‚Äôexpression g√©nique chez chaque souris. Utilisez ensuite la
fonction <code>pivot_longer()</code> pour restaurer le tableau au format
long.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb100">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna1</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span><span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">mouse</span>, <span class="va">expression</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span><span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">mouse</span>, values_from <span class="op">=</span> <span class="va">expression</span><span class="op">)</span></span>
<span><span class="va">rna1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 √ó 23
   gene     `14`    `9`  `10`  `15`  `18`   `6`   `5`  `11`  `22`  `13`  `23`
   &lt;chr&gt;   &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 Asl      1170    361   400   586   626   988   836   535   586   597   938
 2 Apod    36194  10347  9173 10620 13021 29594 24959 13668 13230 15868 27769
 3 Cyp2d22  4060   1616  1603  1901  2171  3349  3122  2008  2254  2277  2985
 4 Klk6      287    629   641   578   448   195   186  1101   537   567   327
 5 Fcrls      85    233   244   237   180    38    68   375   199   177    89
 6 Slc2a4    782    231   248   265   313   786   528   249   266   357   654
 7 Exd2     1619   2288  2235  2513  2366  1359  1474  3126  2379  2173  1531
 8 Gjc2      288    595   568   551   310   146   186   791   454   370   240
 9 Plp1    43217 101241 96534 58354 53126 27173 28728 98658 61356 61647 38019
10 Gnb4     1071   1791  1867  1430  1355   798   806  2437  1394  1554   960
# ‚Ñπ 1,464 more rows
# ‚Ñπ 11 more variables: `24` &lt;dbl&gt;, `8` &lt;dbl&gt;, `7` &lt;dbl&gt;, `1` &lt;dbl&gt;, `16` &lt;dbl&gt;,
#   `21` &lt;dbl&gt;, `4` &lt;dbl&gt;, `2` &lt;dbl&gt;, `20` &lt;dbl&gt;, `12` &lt;dbl&gt;, `19` &lt;dbl&gt;</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb102">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna1</span> <span class="op">%&gt;%</span></span>
<span><span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"mouse_id"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span>, <span class="op">-</span><span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 32,428 √ó 3
   gene  mouse_id counts
   &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt;
 1 Asl   14         1170
 2 Asl   9           361
 3 Asl   10          400
 4 Asl   15          586
 5 Asl   18          626
 6 Asl   6           988
 7 Asl   5           836
 8 Asl   11          535
 9 Asl   22          586
10 Asl   13          597
# ‚Ñπ 32,418 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="question-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="question-1" class="callout-inner">
<h3 class="callout-title">Question</h3>
<div class="callout-content">
<p>Extraire les g√®nes situ√©s sur les chromosomes X et Y du data frame
<code>rna</code> et remodeler les donn√©es avec la variable
<code>sex</code> en colonnes et la varibale <code>chromosome_name</code>
en lignes et l‚Äôexpression moyenne des g√®nes localis√©s dans chaque
chromosome comme valeurs, comme dans le tableau suivant¬†:</p>
<figure><img src="../fig/Exercise_pivot_W.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Il faudra r√©sumer avant de remodeler !</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<p>Calculons d‚Äôabord le niveau d‚Äôexpression moyen des g√®nes des
chromosomes X et Y pour les √©chantillons m√¢les et femelles‚Ä¶</p>
<div class="codewrapper sourceCode" id="cb104">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">arn</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"Y"</span> <span class="op">|</span> <span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"X"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sex</span>, <span class="va">chromosome_name</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">r√©sum√©</span><span class="op">(</span>moyenne <span class="op">=</span> <span class="fu">moyenne</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in r√©sum√©(., moyenne = moyenne(expression)): could not find function "r√©sum√©"</code></pre>
</div>
<p>Et pivotons le tableau au format large</p>
<div class="codewrapper sourceCode" id="cb106">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_1</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"Y"</span> <span class="op">|</span> <span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"X"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sex</span>, <span class="va">chromosome_name</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean <span class="op">=</span> <span class="fu">moyenne</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sexe</span>,</span>
<span>              valeurs_from <span class="op">=</span> <span class="va">moyenne</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `summarise()`:
‚Ñπ In argument: `mean = moyenne(expression)`.
‚Ñπ In group 1: `sex = "Female"` `chromosome_name = "X"`.
Caused by error in `moyenne()`:
! could not find function "moyenne"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb108">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'rna_1' not found</code></pre>
</div>
<p>Maintenant, prenez cette trame de donn√©es et transformez-la avec
<code>pivot_longer()</code> afin que chaque ligne soit une combinaison
unique de <code>chromosome_name</code> par <code>gender</code>.</p>
<div class="codewrapper sourceCode" id="cb110">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_1</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"gender"</span>,</span>
<span>               valeurs_to <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>               <span class="op">-</span><span class="va">chromosome_name</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'rna_1' not found</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `pivot_longer()`:
! Arguments in `...` must be used.
‚úñ Problematic argument:
‚Ä¢ valeurs_to = "mean"
‚Ñπ Did you misspell an argument name?</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="question-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="question-2" class="callout-inner">
<h3 class="callout-title">Question</h3>
<div class="callout-content">
<p>Utilisez l‚Äôensemble de donn√©es <code>rna</code> pour cr√©er une
matrice d‚Äôexpression o√π chaque ligne repr√©sente les niveaux d‚Äôexpression
moyens des g√®nes et les colonnes repr√©sentent les diff√©rents temps.</p>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" aria-labelledby="headingSolution7" data-bs-parent="#accordionSolution7">
<div class="accordion-body">
<p>Calculons d‚Äôabord l‚Äôexpression moyenne par g√®ne et par temps</p>
<div class="codewrapper sourceCode" id="cb113">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">arn</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">g√®ne</span>, <span class="va">temps</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">r√©sum√©</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">moyenne</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in r√©sum√©(., mean_exp = moyenne(expression)): could not find function "r√©sum√©"</code></pre>
</div>
<p>avant d‚Äôutiliser la fonction <code>pivot_wider()</code></p>
<div class="codewrapper sourceCode" id="cb115">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">time</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb117">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 √ó 4
# Groups:   gene [1,474]
   gene         `0`     `4`     `8`
   &lt;chr&gt;      &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
 1 AI504432 1034.   1104.   1014
 2 AW046200  155.    152.     81
 3 AW551984  238     302.    342.
 4 Aamp     4603.   4870    4763.
 5 Abca12      5.29    4.25    4.14
 6 Abcc8    2576.   2609.   2292.
 7 Abhd14a   591.    547.    432.
 8 Abi2     4881.   4903.   4945.
 9 Abi3bp   1175.   1061.    762.
10 Abl2     2170.   2078.   2131.
# ‚Ñπ 1,464 more rows</code></pre>
</div>
<p>Notez que cela g√©n√®re un tibble avec certains noms de colonnes
commen√ßant par un nombre. Si nous voulions s√©lectionner la colonne
correspondant aux points temporels, nous ne pourrions pas utiliser
directement les noms des colonnes‚Ä¶ Que se passe-t-il lorsque l‚Äôon
s√©lectionne la colonne 4¬†?</p>
<div class="codewrapper sourceCode" id="cb119">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">time</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="fl">4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 √ó 2
# Groups:   gene [1,474]
   gene         `8`
   &lt;chr&gt;      &lt;dbl&gt;
 1 AI504432 1014
 2 AW046200   81
 3 AW551984  342.
 4 Aamp     4763.
 5 Abca12      4.14
 6 Abcc8    2292.
 7 Abhd14a   432.
 8 Abi2     4945.
 9 Abi3bp    762.
10 Abl2     2131.
# ‚Ñπ 1,464 more rows</code></pre>
</div>
<p>Pour s√©lectionner le temp 4, il faudrait citer le nom de la colonne,
avec des <em>backticks</em> ‚Äú\`‚Äù</p>
<div class="codewrapper sourceCode" id="cb122">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">time</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">`4`</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 √ó 2
# Groups:   gene [1,474]
   gene         `4`
   &lt;chr&gt;      &lt;dbl&gt;
 1 AI504432 1104.
 2 AW046200  152.
 3 AW551984  302.
 4 Aamp     4870
 5 Abca12      4.25
 6 Abcc8    2609.
 7 Abhd14a   547.
 8 Abi2     4903.
 9 Abi3bp   1061.
10 Abl2     2078.
# ‚Ñπ 1,464 more rows</code></pre>
</div>
<p>Une autre possibilit√© serait de renommer les colonnes, en choisissant
des noms qui ne commencent pas par un chiffre :</p>
<div class="codewrapper sourceCode" id="cb125">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">time</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">rename</span><span class="op">(</span><span class="st">"time0"</span> <span class="op">=</span> <span class="va">`0`</span>, <span class="st">"time4"</span> <span class="op">=</span> <span class="va">`4`</span>, <span class="st">"time8"</span> <span class="op">=</span> <span class="va">`8`</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'gene'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,474 √ó 2
# Groups:   gene [1,474]
   gene       time4
   &lt;chr&gt;      &lt;dbl&gt;
 1 AI504432 1104.
 2 AW046200  152.
 3 AW551984  302.
 4 Aamp     4870
 5 Abca12      4.25
 6 Abcc8    2609.
 7 Abhd14a   547.
 8 Abi2     4903.
 9 Abi3bp   1061.
10 Abl2     2078.
# ‚Ñπ 1,464 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="question-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="question-3" class="callout-inner">
<h3 class="callout-title">Question</h3>
<div class="callout-content">
<p>Utilisez le data frame pr√©c√©dent contenant les niveaux d‚Äôexpression
moyens par temps et cr√©ez une nouvelle colonne contenant les
<em>fold-changes</em> entre les temps 8 et 0, et les temps 8 et 4.
Convertissez ce tableau en un tableau au format long regroupant les
<em>fold-changes</em> calcul√©s.</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" aria-labelledby="headingSolution8" data-bs-parent="#accordionSolution8">
<div class="accordion-body">
<p>√Ä partir du tibble <code>rna_time</code>¬†:</p>
<div class="codewrapper sourceCode" id="cb128">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">arn_time</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'arn_time' not found</code></pre>
</div>
<p>Calculer les <em>fold-changes</em>¬†:</p>
<div class="codewrapper sourceCode" id="cb130">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">muter</span> <span class="op">(</span>time_8_vs_0 <span class="op">=</span> <span class="va">`8`</span> <span class="op">/</span> <span class="va">`0`</span>, time_8_vs_4 <span class="op">=</span> <span class="va">`8`</span> <span class="op">/</span> <span class="va">`4`</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in muter(., time_8_vs_0 = `8`/`0`, time_8_vs_4 = `8`/`4`): could not find function "muter"</code></pre>
</div>
<p>Et utilisez la fonction <code>pivot_longer()</code>¬†:</p>
<div class="codewrapper sourceCode" id="cb132">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>time_8_vs_0 <span class="op">=</span> <span class="va">`8`</span> <span class="op">/</span> <span class="va">`0`</span>, time_8_vs_4 <span class="op">=</span> <span class="va">`8`</span> <span class="op">/</span> <span class="va">`4`</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"comparisons"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"Fold_changes"</span>,</span>
<span>               <span class="va">time_8_vs_0</span><span class="op">:</span><span class="va">time_8_vs_4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 2,948 √ó 6
# Groups:   gene [1,474]
   gene         `0`     `4`     `8` comparisons Fold_changes
   &lt;chr&gt;      &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;              &lt;dbl&gt;
 1 AI504432 1034.   1104.   1014    time_8_vs_0        0.981
 2 AI504432 1034.   1104.   1014    time_8_vs_4        0.918
 3 AW046200  155.    152.     81    time_8_vs_0        0.522
 4 AW046200  155.    152.     81    time_8_vs_4        0.532
 5 AW551984  238     302.    342.   time_8_vs_0        1.44
 6 AW551984  238     302.    342.   time_8_vs_4        1.13
 7 Aamp     4603.   4870    4763.   time_8_vs_0        1.03
 8 Aamp     4603.   4870    4763.   time_8_vs_4        0.978
 9 Abca12      5.29    4.25    4.14 time_8_vs_0        0.784
10 Abca12      5.29    4.25    4.14 time_8_vs_4        0.975
# ‚Ñπ 2,938 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="jointures-de-tables">Jointures de tables<a class="anchor" aria-label="anchor" href="#jointures-de-tables"></a></h2>
<hr class="half-width"><p>Dans de nombreuses situations r√©elles, les donn√©es sont r√©parties sur
plusieurs tables. Cela se produit g√©n√©ralement parce que diff√©rents
types d‚Äôinformations sont collect√©s √† partir de diff√©rentes sources.</p>
<p>Il peut √™tre souhaitable que certaines analyses combinent les donn√©es
de deux ou plusieurs tables en un seul data frame bas√© sur une colonne
commune √† toutes les tables.</p>
<p>Le package <code>dplyr</code> fournit un ensemble de fonctions de
jointure pour combiner deux data frames en fonction des correspondances
entre les colonnes sp√©cifi√©es. Ici, nous fournissons une br√®ve
introduction aux jointures. Pour en savoir plus, veuillez vous r√©f√©rer
au chapitre sur les <a href="https://uclouvain-cbio.github.io/WSBIM1207/sec-join.html" class="external-link">jointures
de table</a>. Le Data Transformation Cheat Sheet fournit √©galement un
bref aper√ßu sur les jointures de table.</p>
<p>Nous allons illustrer la jointure en utilisant une petite table,
<code>rna_mini</code> que nous allons cr√©er en extrayant de la table
<code>rna</code> d‚Äôorigine 3 colonnes et 10 lignes.</p>
<div class="codewrapper sourceCode" id="cb134">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_mini</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>   <span class="fu">select</span><span class="op">(</span><span class="va">g√®ne</span>, <span class="va">√©chantillon</span>, <span class="va">expression</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>   <span class="fu">head</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `select()`:
! Can't select columns that don't exist.
‚úñ Column `g√®ne` doesn't exist.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb136">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_mini</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'rna_mini' not found</code></pre>
</div>
<p>Le deuxi√®me tableau, <code>annot1</code>, contient 2 colonnes,
<code>gene</code> et <code>gene_description</code>. Vous pouvez soit <a href="https://carpentries-incubator.github.io/bioc-intro/data/annot1.csv" class="external-link">t√©l√©charger
annot1.csv</a> en cliquant sur le lien puis en d√©pla√ßant le fichier dans
le dossier <code>data/</code>, ou vous pouvez utiliser le code R
ci-dessous pour le t√©l√©charger directement dans le dossier
<code>data</code>.</p>
<div class="codewrapper sourceCode" id="cb138">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://carpentries-incubator.github.io/bioc-intro/data/annot1.csv"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/annot1.csv"</span><span class="op">)</span></span>
<span><span class="va">annot1</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/annot1.csv"</span><span class="op">)</span></span>
<span><span class="va">annot1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 10 √ó 2
   gene    gene_description
   &lt;chr&gt;   &lt;chr&gt;
 1 Cyp2d22 cytochrome P450, family 2, subfamily d, polypeptide 22 [Source:MGI S‚Ä¶
 2 Klk6    kallikrein related-peptidase 6 [Source:MGI Symbol;Acc:MGI:1343166]
 3 Fcrls   Fc receptor-like S, scavenger receptor [Source:MGI Symbol;Acc:MGI:19‚Ä¶
 4 Plp1    proteolipid protein (myelin) 1 [Source:MGI Symbol;Acc:MGI:97623]
 5 Exd2    exonuclease 3'-5' domain containing 2 [Source:MGI Symbol;Acc:MGI:192‚Ä¶
 6 Apod    apolipoprotein D [Source:MGI Symbol;Acc:MGI:88056]
 7 Gnb4    guanine nucleotide binding protein (G protein), beta 4 [Source:MGI S‚Ä¶
 8 Slc2a4  solute carrier family 2 (facilitated glucose transporter), member 4 ‚Ä¶
 9 Asl     argininosuccinate lyase [Source:MGI Symbol;Acc:MGI:88084]
10 Gjc2    gap junction protein, gamma 2 [Source:MGI Symbol;Acc:MGI:2153060]    </code></pre>
</div>
<p>Nous voulons maintenant joindre ces deux tables en une seule
contenant toutes les variables en utilisant la fonction
<code>full_join()</code> du package <code>dplyr</code>. La fonction
trouvera automatiquement la variable commune correspondant aux colonnes
de la premi√®re et de la deuxi√®me table. Dans ce cas, <code>gene</code>
est la seule variable commune. De telles variables sont appel√©es cl√©s.
Les cl√©s sont utilis√©es pour faire correspondre les observations dans
diff√©rentes tables.</p>
<div class="codewrapper sourceCode" id="cb140">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">full_join</span><span class="op">(</span><span class="va">rna_mini</span>, <span class="va">annot1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'rna_mini' not found</code></pre>
</div>
<p>En recherche, les g√®nes sont parfois nom√©s diff√©remment.</p>
<p>La table <code>annot2</code> est exactement la m√™me que
<code>annot1</code> sauf que la variable contenant les noms de g√®nes est
nomm√©e diff√©remment. Encore une fois, <a href="https://carpentries-incubator.github.io/bioc-intro/data/annot2.csv" class="external-link">t√©l√©chargez
annot2.csv</a> vous-m√™me et d√©placez-le vers <code>data/</code> ou
utilisez le code R ci-dessous.</p>
<div class="codewrapper sourceCode" id="cb142">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://carpentries-incubator.github.io/bioc-intro/data/annot2.csv"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/annot2.csv"</span><span class="op">)</span></span>
<span><span class="va">annot2</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/annot2.csv"</span><span class="op">)</span></span>
<span><span class="va">annot2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 10 √ó 2
   external_gene_name description
   &lt;chr&gt;              &lt;chr&gt;
 1 Cyp2d22            cytochrome P450, family 2, subfamily d, polypeptide 22 [S‚Ä¶
 2 Klk6               kallikrein related-peptidase 6 [Source:MGI Symbol;Acc:MGI‚Ä¶
 3 Fcrls              Fc receptor-like S, scavenger receptor [Source:MGI Symbol‚Ä¶
 4 Plp1               proteolipid protein (myelin) 1 [Source:MGI Symbol;Acc:MGI‚Ä¶
 5 Exd2               exonuclease 3'-5' domain containing 2 [Source:MGI Symbol;‚Ä¶
 6 Apod               apolipoprotein D [Source:MGI Symbol;Acc:MGI:88056]
 7 Gnb4               guanine nucleotide binding protein (G protein), beta 4 [S‚Ä¶
 8 Slc2a4             solute carrier family 2 (facilitated glucose transporter)‚Ä¶
 9 Asl                argininosuccinate lyase [Source:MGI Symbol;Acc:MGI:88084]
10 Gjc2               gap junction protein, gamma 2 [Source:MGI Symbol;Acc:MGI:‚Ä¶</code></pre>
</div>
<p>Si aucun des noms de variables ne correspond, nous pouvons d√©finir
manuellement les variables √† utiliser pour la correspondance. Ces
variables peuvent √™tre d√©finies en utilisant l‚Äôargument <code>by</code>,
comme indiqu√© ci-dessous avec les tables <code>rna_mini</code> et
<code>annot2</code>.</p>
<div class="codewrapper sourceCode" id="cb144">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">full_join</span><span class="op">(</span><span class="va">rna_mini</span>, <span class="va">annot2</span>, by <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene"</span> <span class="op">=</span> <span class="st">"external_gene_name"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'rna_mini' not found</code></pre>
</div>
<p>Comme on peut le voir ci-dessus, le nom de variable de la premi√®re
table est lors de la jointure.</p>
<div id="d√©fi-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="d√©fi-4" class="callout-inner">
<h3 class="callout-title">D√©fi:</h3>
<div class="callout-content">
<p>T√©l√©chargez la table <code>annot3</code> en cliquant <a href="https://carpentries-incubator.github.io/bioc-intro/data/annot3.csv" class="external-link">ici</a>
et placez la table dans votre dossier <code>data</code>. √Ä l‚Äôaide de la
fonction <code>full_join()</code>, joignez les tables
<code>rna_mini</code> et <code>annot3</code>. Que s‚Äôest-il pass√© pour
les g√®nes <em>Klk6</em>, <em>mt-Tf</em>, <em>mt-Rnr1</em>,
<em>mt-Tv</em>, <em>mt-Rnr2</em> et <em>mt-Tl1</em> ?</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" aria-labelledby="headingSolution9" data-bs-parent="#accordionSolution9">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb146">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annot3</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/annot3.csv"</span><span class="op">)</span></span>
<span><span class="fu">full_join</span><span class="op">(</span><span class="va">rna_mini</span>, <span class="va">annot3</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'rna_mini' not found</code></pre>
</div>
<p>Les g√®nes <em>Klk6</em> ne sont pr√©sents que dans
<code>rna_mini</code>, tandis que les g√®nes <em>mt-Tf</em>,
<em>mt-Rnr1</em>, <em>mt-Tv</em>, <em>mt-Rnr2</em> et <em>mt-Tl1</em>
sont pr√©sent uniquement dans la table <code>annot3</code>. Leurs valeurs
respectives pour les variables du tableau ont √©t√© cod√©es comme
manquantes.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="exporter-des-donn√©es">Exporter des donn√©es<a class="anchor" aria-label="anchor" href="#exporter-des-donn%C3%A9es"></a></h2>
<hr class="half-width"><p>Maintenant que vous avez appris √† utiliser <code>dplyr</code> pour
extraire des informations de ou r√©sumer vos donn√©es, vous souhaiterez
peut-√™tre exporter ces nouveaux jeux de donn√©es pour les partager avec
vos collaborateurs ou pour les archiver.</p>
<p>Tout comme √† la fonction <code>read_csv()</code> utilis√©e pour
lire/importer les fichiers CSV dans R, il existe une fonction
<code>write_csv()</code> qui g√©n√®re des fichiers CSV √† partir de data
frames.</p>
<p>Avant d‚Äôutiliser <code>write_csv()</code>, nous allons cr√©er un
nouveau dossier, <code>data_output</code>, dans notre r√©pertoire de
travail, pour y sauvegarder nos nouveaux de donn√©es. Nous ne voulons pas
√©crive les ensembles de donn√©es g√©n√©r√©s dans le m√™me r√©pertoire que nos
donn√©es brutes. C‚Äôest en effet une bonne pratique de les garder s√©par√©s.
Le dossier <code>data</code> ne doit contenir que les donn√©es brutes et
non modifi√©es, et doit √™tre laiss√© tel quel pour nous assurer que nous
ne les supprimons ou modifions pas. En revanche, notre script g√©n√©rera
le contenu du r√©pertoire <code>data_output</code>, donc m√™me si les
fichiers qu‚Äôil contient sont supprim√©s, nous pourrons toujours les
reg√©n√©rer.</p>
<p>Utilisons <code>write_csv()</code> pour sauvegarder la table
<code>rna_wide</code> que nous avons cr√©√©e pr√©c√©demment.</p>
<div class="codewrapper sourceCode" id="cb148">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">write_csv</span><span class="op">(</span><span class="va">rna_wide</span>, file <span class="op">=</span> <span class="st">"data_output/rna_wide.csv"</span><span class="op">)</span></span></code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>Donn√©es tabulaires dans R utilisant le m√©ta-package
<em>tidyverse</em>.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/25-starting-with-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/40-visualization.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/25-starting-with-data.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Commencer par les
        </a>
        <a class="chapter-link float-end" href="../instructor/40-visualization.html" rel="next">
          Next: Visualisation de...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/episodes/30-dplyr.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/bioc-intro/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/bioc-intro/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-intro/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:laurent.gatto@uclouvain.be">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/bioc-intro/instructor/30-dplyr.html",
  "inLanguage": "fr",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Manipulation et analyse de donn√©es avec dplyr",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/bioc-intro/instructor/30-dplyr.html",
  "identifier": "https://carpentries-incubator.github.io/bioc-intro/instructor/30-dplyr.html",
  "dateCreated": "2020-09-14",
  "dateModified": "2025-05-05",
  "datePublished": "2025-05-05"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

